Add B3 propagator

MikeGoldsmith opened the pull request:
Fixes #853.

Adds support for B3 propagation, including both single and multi header formats.

MikeGoldsmith commented:
How do I request a new coverage report?

jkwatson commented on api/src/main/java/io/opentelemetry/trace/propagation/B3Propagator.java:
```diff
@@ -0,0 +1,201 @@
+   *
+   * @param singleHeader whether to use single or multiple headers.
+   */
+  public B3Propagator(Boolean singleHeader) {
```
does this need to be a nullable big-b `Boolean`? Can't it just be a primitive boolean?

MikeGoldsmith replied:
Pushed an update - there is no need for a `Boolean` in this case.

jkwatson commented on api/src/main/java/io/opentelemetry/trace/propagation/B3Propagator.java:
```diff
@@ -0,0 +1,201 @@
+    String traceId;
+    String spanId;
+    String sampled;
+    if (singleHeader) {
```
This should be broken up into at least 2 separate methods. This method is definitely too long. One method for the single header case that returns the context, and one for the multi-header case that returns the context will make this code much easier to read. Thanks!

carlosalberto replied:
I support this. The TraceContext one is refactored as it was too long and it was (relatively) hard to read.

carlosalberto commented:
Hey @MikeGoldsmith thanks for the PR, it looks very neat and clean. Left a small comment supporting @jkwatson's suggestion to split the extraction method.

One thing I remember is that B3 is expected to handle case insensitive headers for extraction, i.e. use either. `X-B3-TraceId` or. `x-b3-traceid`. At least that's what a few vendors support (DDog and LS for sure). 

The second one is about the single/multiple mode. I'd imagine some users wanting to have both modes supported (by trying the single header case first, falling back to the multiple headers case). This would be specially useful in case they interact with other clients that might send single or multiple modes. wWe could do that in a follow up PR if/as needed though ;)

thisthat commented on api/src/main/java/io/opentelemetry/trace/propagation/B3Propagator.java:
```diff
@@ -0,0 +1,201 @@
+  }
+
+  private static boolean isTraceIdValid(String value) {
+    return !(value == null || value.isEmpty() || value.length() > MAX_TRACE_ID_LENGTH);
```
The first two checks could leverage the `StringUtils.isNullOrEmpty(value)` method

MikeGoldsmith replied:
Thanks - I've pushed an update to use `StringUtils`.

MikeGoldsmith commented:
@carlosalberto I've refactored the extract into two separate methods and addressed the other feedback.

Other implementations of B3 are case-insensitive, but I don't believe any of the OTel version are. This is because we are unable to `get` a key regardless of casing, or retrieve all known keys and compare locally. In OpenTracing with ITextMap, it was possible to get all keys and do a local case insensitive lookup.

I've created issue #985 to track making the B3 propagator automatically switch between single and multi headers.

bogdandrutu commented on api/src/main/java/io/opentelemetry/trace/propagation/B3Propagator.java:
```diff
@@ -0,0 +1,204 @@
+              .setIsSampled(
+                  TRUE_INT.equals(sampled)
+                      || Boolean.parseBoolean(sampled)) // accept either "1" or "true"
+              .build();
```
You can construct a sampled/notSampled TraceFlags and reuse them.

bogdandrutu commented on api/src/main/java/io/opentelemetry/trace/propagation/B3Propagator.java:
```diff
@@ -0,0 +1,204 @@
+              .build();
+
+      return SpanContext.createFromRemoteParent(
+          TraceId.fromLowerBase16(traceId, 0),
```
This requires traceId (string) to be 32 hex characters and we only check that the size is not greater than MAX_TRACE_ID_LENGTH.

MikeGoldsmith replied:
Pushed change to left pad TraceIds upto max with leading 0's. I added padLeft to `StringUtils`. Let me know what you think.

carlosalberto replied:
As this was fixed along with a test (see https://github.com/open-telemetry/opentelemetry-java/pull/979/files#diff-004fb9fadd7caedae3844dfe8b040186R137) I'm marking this as resolved.

bogdandrutu commented on api/src/main/java/io/opentelemetry/trace/propagation/B3Propagator.java:
```diff
@@ -0,0 +1,204 @@
+ * href=https://github.com/openzipkin/b3-propagation>openzipkin/b3-propagation</a>.
+ */
+@Immutable
+public class B3Propagator implements HttpTextFormat<SpanContext> {
```
Should we have 2 propagators: legacy + single-header? I think that simplifies the code. We can do it later.

MikeGoldsmith replied:
Created #999 to track breaking single and multiple header implementations.

bogdandrutu commented on api/src/main/java/io/opentelemetry/trace/propagation/B3Propagator.java:
```diff
@@ -0,0 +1,204 @@
+      return SpanContext.getInvalid();
+    }
+
+    String sampled = parts.length == 3 ? parts[2] : null;
```
Based on my understanding sampled is present if parts.length is 3 or 4 :)

MikeGoldsmith replied:
You're correct. We need to check the sampled string if we have three or more items in the array.

MikeGoldsmith commented on api/src/main/java/io/opentelemetry/trace/propagation/B3Propagator.java:
```diff
@@ -0,0 +1,204 @@
+  }
+
+  @Override
+  public <C /*>>> extends @NonNull Object*/> SpanContext extract(C carrier, Getter<C> getter) {
```
Removed :+1:
